### PR TITLE
Allow rpc_user and rpc_pass in config file

### PR DESCRIPTION
**Pull request by raphjaph**
#1399 

Probably closes:
#1785
#1562

### CODE REVIEW

**Comment by raphjaph** [comment]
Let's not support this as flags but as a option in the config file

**Comment by bliotti** [comment]
Probably need to change Title "Instead of"?  
Was worried we were dropping  .cookie

**Comment by bliotti** [comment]
Tested [37d36ea](https://github.com/casey/ord/pull/1527/commits/37d36ea00d0607ed1c52a790423e265bf4345baa)  , both methods work.

**However `rpcuser` and `rpcpassword` are deprecated.** 

```
2023-02-06 21:34:02 bitcoin.conf file exists, skipping.
2023-02-06 21:34:02 2023-02-07T02:34:02Z Bitcoin Core version v24.0.1 (release build)
2023-02-06 21:34:02 2023-02-07T02:34:02Z Signet derived magic (message start): 0a03cf40
2023-02-06 21:34:02 2023-02-07T02:34:02Z Using the 'sse4(1way),sse41(4way),avx2(8way)' SHA256 implementation
2023-02-06 21:34:02 2023-02-07T02:34:02Z Using RdSeed as an additional entropy source
2023-02-06 21:34:02 2023-02-07T02:34:02Z Using RdRand as an additional entropy source
2023-02-06 21:34:02 2023-02-07T02:34:02Z Default data directory /root/.bitcoin
2023-02-06 21:34:02 2023-02-07T02:34:02Z Using data directory /root/.bitcoin/signet
2023-02-06 21:34:02 2023-02-07T02:34:02Z Config file: /root/.bitcoin/bitcoin.conf
2023-02-06 21:34:02 2023-02-07T02:34:02Z Config file arg: rpcpassword=****
2023-02-06 21:34:02 2023-02-07T02:34:02Z Config file arg: rpcuser=****
2023-02-06 21:34:02 2023-02-07T02:34:02Z Config file arg: server="1"
2023-02-06 21:34:02 2023-02-07T02:34:02Z Config file arg: signet="1"
2023-02-06 21:34:02 2023-02-07T02:34:02Z Config file arg: txindex="1"

2023-02-06 21:34:02 2023-02-07T02:34:02Z Config options rpcuser and rpcpassword will soon be deprecated. Locally-run instances may remove rpcuser to use cookie-based auth, or may be replaced with rpcauth. Please see share/rpcauth for rpcauth auth generation.

2023-02-06 21:34:02 2023-02-07T02:34:02Z [http] starting 4 worker threads
2023-02-06 21:34:02 2023-02-07T02:34:02Z Using wallet directory /root/.bitcoin/signet/wallets
2023-02-06 21:34:02 2023-02-07T02:34:02Z init message: Verifying wallet(s)…
2023-02-06 21:34:02 2023-02-07T02:34:02Z Using /16 prefix for IP bucketing
```

**Comment by raphjaph** [comment]
Yea, using rpc_user and rpc_pass should not be encouraged

**Comment by andrewtoth** [comment]
> rpcuser and rpcpassword will soon be deprecated. 

They will not be deprecated from use in clients. The deprecation refers to using bare `rpcuser/rpcpassword` in Bitcoin Core's `bitcoin.conf`. When configuring bitcoind you should not use `rpcuser/rpcpassword` but `rpcauth`. For clients `rpcuser/rpcpassword` will never be deprecated, how else would they authenticate without access to the cookie file?

**Comment by raphjaph** [comment]
> They will not be deprecated from use in _clients_. The deprecation refers to using bare `rpcuser/rpcpassword` in favor of generating an `rpcauth` with the user and password in Bitcoin Core's `bitcoin.conf`. When configuring bitcoind you should not use `rpcuser/rpcpassword` but `rpcauth`, which does not expose the password in plaintext in the config file. For clients `rpcuser/rpcpassword` will not be deprecated.

Yes, true. Passing credentials as command line flags is still not ideal because those will be stored in plaintext in your bash history. I've add it as flags anyway because that fear is probably overblown.

**Comment by raphjaph** [comment]
`rpc_user` and `rpc_pass` can now be passed in to `ord` in a plethora of ways. In order of precedence:
- in the `ord.yaml` config file
- as environmental variables
- as command line options
- through the cookie file

Generally I think we have to overhaul or configuration setup. I think there should be global `Config` or `Options` struct maybe initialized with `lazy_static!` à la [this](https://www.philipdaniels.com/blog/2019/rust-configuration-example/) method. We were creating rpc clients in different places in `index.rs` that didn't go through or `Option` methods. These were verifying that the client is and Bitcoin Core are the correct version, etc. This PR also contains some `.clone()`s in places that might be inefficient.

**Comment by Sjors** [comment]
In addition to these methods, you could also read `rpcuser` and `rpcpassword` from `bitcoin.conf`, like `bitcoin-cli` does. For example a user might want to run their `bitcoind` under a different user than `ord`, in which case it won't have access to the cookie file, but you can create a `bitcoin.conf` file with the RPC credentials. (this isn't super useful in practice because Bitcoin Core doesn't have a way yet to limit RPC permissions)

**Comment by raphjaph** [comment]
> In addition to these methods, you could also read `rpcuser` and `rpcpassword` from `bitcoin.conf`, like `bitcoin-cli` does. For example a user might want to run their `bitcoind` under a different user than `ord`, in which case it won't have access to the cookie file, but you can create a `bitcoin.conf` file with the RPC credentials. (this isn't super useful in practice because Bitcoin Core doesn't have a way yet to limit RPC permissions)

I think `rpcuser/rpcpass` from `bitcoin.conf` is being deprecated in favor of `rpcauth`. This will only store the username, password hash and salt.

**Comment by Sjors** [comment]
Assuming a system with two users: `bitcoin` and `ord`. It makes sense for the `bitcoin` user to use `rpcauth` so only the hash is stored. But the `ord` user would use `rpcpass`. It's a bit confusing that both `bitcoind` and `bitcoin-cli` use the same config file. I don't think this will stop working anytime soon.

File: src/options.rs [hunk 1]
```diff
@@ -137,24 +141,45 @@ impl Options {
...
-
     let rpc_url = self.rpc_url();
 
-    log::info!(
-      "Connecting to Bitcoin Core RPC server at {rpc_url} using credentials from `{}`",
-      cookie_file.display()
-    );
+    let config = self.load_config().unwrap();
+
+    let auth = if let (Some(rpc_user), Some(rpc_pass)) = (config.rpc_user, config.rpc_pass) {
```
Comment (casey) [inline src/options.rs:148]: For context:

https://stackoverflow.com/questions/32272911/precedence-of-configuration-options-environment-registry-configuration-file-a

File: src/options.rs [hunk 2]
```diff
@@ -137,24 +141,45 @@ impl Options {
...
+      Auth::UserPass(rpc_user, rpc_pass)
+    } else if let (Some(rpc_user), Some(rpc_pass)) = (self.rpc_user.clone(), self.rpc_pass.clone())
+    {
+      log::info!(
+        "Connecting to Bitcoin Core RPC server at {rpc_url} using rpc_pass for `{rpc_user}`",
+      );
+
+      Auth::UserPass(rpc_user, rpc_pass)
+    } else if let (Some(rpc_user), Some(rpc_pass)) =
+      (env::var_os("RPC_USER"), env::var_os("RPC_PASS"))
```
Comment (casey) [inline src/options.rs:162]: disaggregate `rpc_user` and `rpc_pass`

File: src/options.rs [hunk 3]
```diff
@@ -561,6 +586,45 @@ mod tests {
         .unwrap(),
       Config {
         hidden: iter::once(id).collect(),
+        rpc_user: None,
+        rpc_pass: None,
+      }
+    );
+  }
+
+  #[test]
```
Comment (casey) [inline src/options.rs:595]: unit test and precedent test

File: src/options.rs [hunk 4]
```diff
@@ -561,6 +586,45 @@ mod tests {
...
+  fn test_rpc_user_and_pass_flags() {
+    let options = Arguments::try_parse_from([
+      "ord",
+      "--rpc-user",
+      "satoshi",
```
Comment (casey) [inline src/options.rs:600]: use `foo` and `bar`

**Review by casey — CHANGES_REQUESTED** [review]
Great start! Let's do the following:

- Write a function which takes a cli flag value, a environment variable, a config file option, and a default value, and returns the correct value. Unit test this so we don't have to test all combinations of configuration layers in integration tests.
- Write integration tests which test all three configuration layers independently: Config with CLI flags, environment variables, config file, and default value for RPC user and RPC pass. (We don't need to comprehensively test existing options.)
- Use `foo`, `bar`, and `baz` consistently when arbitrary strings or variable names are needed in tests. (Don't need to refactor other stuff, just this PR.

File: src/index/updater.rs [hunk 5]
```diff
@@ -34,6 +34,7 @@ pub(crate) struct Updater {
   height: u64,
   index_sats: bool,
   sat_ranges_since_flush: u64,
+  options: Options,
```
Comment (andrewtoth) [inline src/index/updater.rs:37]: Do we really need to make `options`  a member of the `Updater` struct? It's only used one time on line 101, `let rx = Self::fetch_blocks_from(index, self.height, self.index_sats, self.options.clone())?;` where we are already passing in `index` which has the `options` as a member already.

Comment (raphjaph) [reply src/index/updater.rs:37]: good point, I'll try to remove it.

File: src/index/updater.rs [hunk 6]
```diff
@@ -179,13 +179,13 @@ impl Updater {
...
     index_sats: bool,
+    options: Options,
   ) -> Result<mpsc::Receiver<BlockData>> {
     let (tx, rx) = mpsc::sync_channel(32);
 
     let height_limit = index.height_limit;
 
-    let client =
-      Client::new(&index.rpc_url, index.auth.clone()).context("failed to connect to RPC URL")?;
+    let client = options.bitcoin_rpc_client()?;
```
Comment (andrewtoth) [inline src/index/updater.rs:188]: Why not just `let client = index.options.bitcoin_rpc_client()?;`, and then we don't have to pass the cloned `options` in here at all?

File: src/options.rs [hunk 7]
```diff
@@ -136,25 +140,56 @@ impl Options {
...
+    log::info!("Connecting to Bitcoin Core at {}", self.rpc_url());
 
-    log::info!(
-      "Connecting to Bitcoin Core RPC server at {rpc_url} using credentials from `{}`",
-      cookie_file.display()
+    let rpc_user = Options::derive_var(
+      self.rpc_user.clone().map(|string| string.into()),
+      env::var_os("RPC_USER"),
+      self
+        .load_config()
```
Comment (andrewtoth) [inline src/options.rs:159]: Could load config and store in a variable above this statement and reuse below for `rpc_pass`?

File: src/options.rs [hunk 8]
```diff
@@ -136,25 +140,56 @@ impl Options {
...
 
-    let rpc_url = self.rpc_url();
+  pub(crate) fn auth(&self) -> Auth {
+    log::info!("Connecting to Bitcoin Core at {}", self.rpc_url());
 
-    log::info!(
-      "Connecting to Bitcoin Core RPC server at {rpc_url} using credentials from `{}`",
-      cookie_file.display()
+    let rpc_user = Options::derive_var(
+      self.rpc_user.clone().map(|string| string.into()),
```
Comment (andrewtoth) [inline src/options.rs:156]:
```suggestion
      self.rpc_user.as_ref().map(|string| string.into()),
```

File: src/options.rs [hunk 9]
```diff
@@ -136,25 +140,56 @@ impl Options {
...
 
-    let client =
-      Client::new(&rpc_url, Auth::CookieFile(cookie_file.clone())).with_context(|| {
-        format!(
-          "failed to connect to Bitcoin Core RPC at {rpc_url} using cookie file {}",
-          cookie_file.display()
-        )
-      })?;
+    let rpc_pass = Options::derive_var(
+      self.rpc_pass.clone().map(|string| string.into()),
```
Comment (andrewtoth) [inline src/options.rs:167]:
```suggestion
      self.rpc_pass.as_ref().map(|string| string.into()),
```

File: src/index/updater.rs (hunk 6 above)
Comment (raphjaph) [reply src/index/updater.rs:188]: jup

Comment (raphjaph) [reply src/index/updater.rs:188]: jup

**Comment by raphjaph** [comment]
Thanks for the review @andrewtoth! 

I've updated the things.

File: src/subcommand/server.rs [hunk 10]
```diff
@@ -2266,7 +2288,8 @@ mod tests {
 
   #[test]
   fn feed() {
-    let server = TestServer::new_with_sat_index();
+    // let server = TestServer::new_with_sat_index();
```
Comment (andrewtoth) [inline src/subcommand/server.rs:2291]: Commented line should be removed

File: src/subcommand/server.rs [hunk 11]
```diff
@@ -937,6 +937,28 @@ mod tests {
       Self::new_server(test_bitcoincore_rpc::spawn(), None, ord_args, server_args)
     }
 
+    fn new_with_regtest() -> Self {
```
Comment (andrewtoth) [inline src/subcommand/server.rs:940]:
```suggestion
    fn new_regtest() -> Self {
```
seems simpler?

**Review by andrewtoth — CHANGES_REQUESTED** [review]

**Comment by andrewtoth** [comment]
Two small nits but otherwise LGTM.

File: src/config.rs [hunk 12]
```diff
@@ -3,6 +3,8 @@ use super::*;
 #[derive(Deserialize, Default, PartialEq, Debug)]
 pub(crate) struct Config {
   pub(crate) hidden: HashSet<InscriptionId>,
+  pub(crate) rpc_user: Option<String>,
+  pub(crate) rpc_pass: Option<String>,
```
Comment (casey) [inline src/config.rs:7]: Struct fields should be sorted.

File: src/options.rs [hunk 13]
```diff
@@ -37,6 +37,10 @@ pub(crate) struct Options {
   pub(crate) index_sats: bool,
   #[clap(long, short, help = "Use regtest. Equivalent to `--chain regtest`.")]
   pub(crate) regtest: bool,
+  #[clap(long, help = "Authenticate to Bitcoin Core RPC as <RPC_USER>.")]
+  pub(crate) rpc_user: Option<String>,
+  #[clap(long, help = "Authenticate to Bitcoin Core RPC with <RPC_PASS>.")]
+  pub(crate) rpc_pass: Option<String>,
```
Comment (casey) [inline src/options.rs:43]: I notice that both `rpc_user` and `rpc_pass` are optional. What is the behavior when we pass `rpc_user` without `rpc_pass`, and vice versa? We should have test coverage for all four cases. (up/Up/uP/UP)

File: src/options.rs [hunk 14]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
-    let cookie_file = self
-      .cookie_file()
-      .map_err(|err| anyhow!("failed to get cookie file path: {err}"))?;
+  fn derive_var(
+    arg: Option<OsString>,
+    env: Option<OsString>,
+    config: Option<OsString>,
+    default: Option<OsString>,
+  ) -> Option<OsString> {
+    arg.or(env).or(config).or(default)
```
Comment (casey) [inline src/options.rs:149]: Genau!

File: src/options.rs [hunk 15]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
+  }
 
-    let rpc_url = self.rpc_url();
+  pub(crate) fn auth(&self) -> Auth {
+    log::info!("Connecting to Bitcoin Core at {}", self.rpc_url());
+
+    let config = self.load_config().unwrap();
```
Comment (casey) [inline src/options.rs:155]: This function must return an error, and we should include here a good error message. There are many reasons why config loading might fail, so this will be a common error that users encounter.

File: src/options.rs [hunk 16]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
 
-  pub(crate) fn bitcoin_rpc_client(&self) -> Result<Client> {
-    let cookie_file = self
-      .cookie_file()
-      .map_err(|err| anyhow!("failed to get cookie file path: {err}"))?;
+  fn derive_var(
+    arg: Option<OsString>,
+    env: Option<OsString>,
+    config: Option<OsString>,
+    default: Option<OsString>,
```
Comment (casey) [inline src/options.rs:147]: The current function signature doesn't allow passing, for example, strings, osstr, cstr, etc. We should make the function signature more generic, so the caller can pass any combination of types that can be converted into an osstr, which will allow fewer conversions at the call site.

I was thinking of suggesting this:

```suggestion
    arg: Option<impl AsRef<OsStr>>,
    env: Option<impl AsRef<OsStr>>,
    config: Option<impl AsRef<OsStr>>,
    default: Option<impl AsRef<OsStr>>,
```

But use of generics forces the compiler to emit a monomerized function for every combination of argument types.

I'm leaning towards this, which should be faster to compile, and which is a bit simpler to read:

```suggestion
    arg: Option<&OsStr>,
    env: Option<&OsStr>,
    config: Option<&OsStr>,
    default: Option<&OsStr>,
```

File: src/options.rs [hunk 17]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
-    let rpc_url = self.rpc_url();
+  pub(crate) fn auth(&self) -> Auth {
+    log::info!("Connecting to Bitcoin Core at {}", self.rpc_url());
+
+    let config = self.load_config().unwrap();
+
+    let rpc_user = Options::derive_var(
+      self.rpc_user.as_ref().map(|string| string.into()),
+      env::var_os("RPC_USER"),
+      config.rpc_user.map(|string| string.into()),
```
Comment (casey) [inline src/options.rs:160]: This will need some fiddling to work with the changed function signature.

File: src/options.rs [hunk 18]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
+      None,
+    );
 
-    log::info!(
-      "Connecting to Bitcoin Core RPC server at {rpc_url} using credentials from `{}`",
-      cookie_file.display()
+    let rpc_pass = Options::derive_var(
+      self.rpc_pass.as_ref().map(|string| string.into()),
```
Comment (casey) [inline src/options.rs:165]: This will need some fiddling to work with the changed function signature.

**Review by casey — CHANGES_REQUESTED** [review]
Did a partial review, check out my comments!

File: src/options.rs [hunk 19]
```diff
@@ -136,25 +140,50 @@ impl Options {
     )
   }
 
-  pub(crate) fn bitcoin_rpc_client(&self) -> Result<Client> {
-    let cookie_file = self
-      .cookie_file()
-      .map_err(|err| anyhow!("failed to get cookie file path: {err}"))?;
+  fn derive_var(
```
Comment (casey) [inline src/options.rs:143]: Also, drive-by post-partial-review comment: I think this function should take `Option<&str>` and return `Option<&str>`.

File: src/options.rs [hunk 20]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
-    let client =
-      Client::new(&rpc_url, Auth::CookieFile(cookie_file.clone())).with_context(|| {
-        format!(
-          "failed to connect to Bitcoin Core RPC at {rpc_url} using cookie file {}",
-          cookie_file.display()
-        )
-      })?;
+    match (rpc_user, rpc_pass) {
+      (Some(rpc_user), Some(rpc_pass)) => Auth::UserPass(
+        rpc_user.into_string().expect("rpc_user is invalid UTF-8"),
```
Comment (casey) [inline src/options.rs:173]: These should be converted to `&str` at the call site. Additionally, we need to return an error if the rpc_user and rpc_pass aren't UTF-8

File: src/options.rs [hunk 21]
```diff
@@ -136,25 +140,50 @@ impl Options {
...
+        rpc_pass.into_string().expect("rpc_pass is invalid UTF-8"),
+      ),
+      _ => Auth::CookieFile(self.cookie_file().unwrap()),
```
Comment (casey) [inline src/options.rs:176]: Unwraps should be avoided. Also, the cookie file not being present is a common error, so we need to return a good error message.

File: src/options.rs [hunk 22]
```diff
@@ -561,6 +590,39 @@ mod tests {
         .unwrap(),
       Config {
         hidden: iter::once(id).collect(),
+        rpc_user: None,
+        rpc_pass: None,
```
Comment (casey) [inline src/options.rs:594]: These should be sorted.

**Review by casey — CHANGES_REQUESTED** [review]
Some more changes.

File: src/options.rs (hunk 13 above)
Comment (raphjaph) [reply src/options.rs:43]: It tries to get it from env or config file. Finally it looks at the cookie file.

File: src/options.rs [hunk 23]
```diff
@@ -136,25 +140,69 @@ impl Options {
...
+
+    let rpc_pass = Options::derive_var(
+      self.bitcoin_rpc_pass.as_deref(),
+      Some("BITCOIN_RPC_PASS"),
+      config.rpc_pass.as_deref(),
+      None,
+    )?;
+
+    match (rpc_user, rpc_pass) {
+      (Some(rpc_user), Some(rpc_pass)) => Ok(Auth::UserPass(rpc_user, rpc_pass)),
```
Comment (casey) [inline src/options.rs:185]: We should handle the `None/Some` and `Some/None` cases by throwing an error, and we should have tests for both of these cases.

File: tests/index.rs [hunk 24]
```diff
@@ -1,4 +1,4 @@
-use super::*;
+use {super::*, crate::command_builder::ToArgs};
```
Comment (casey) [inline tests/index.rs:1]: What is the import used for? Can we remove it?

File: tests/index.rs [hunk 25]
```diff
@@ -35,3 +35,32 @@ fn re_opening_database_does_not_trigger_schema_check() {
     .rpc_server(&rpc_server)
     .run();
 }
+
+#[test]
+fn index_runs_with_rpc_user_and_pass_as_env_vars() {
```
Comment (casey) [inline tests/index.rs:40]: Let's modify this test to make sure we're using the correct values. Let's record the user and pass which are passed each time a client connects, and then do something like `assert_eq!(rpc_server.client_credentials, &[("foo", "bar")])`.

File: tests/server.rs [hunk 26]
```diff
@@ -298,3 +298,50 @@ fn expected_sat_time_is_rounded() {
     r".*<dt>timestamp</dt><dd><time>.* \d+:\d+:\d+ UTC</time> \(expected\)</dd>.*",
   );
 }
+
+#[test]
+fn server_runs_with_rpc_user_and_pass_as_env_vars() {
```
Comment (casey) [inline tests/server.rs:303]: Let's do the same style of test in index_runs_with_rpc_user_and_pass_as_env_vars

**Review by casey — CHANGES_REQUESTED** [review]
SO CLOSE

File: tests/index.rs (hunk 24 above)
Comment (raphjaph) [reply tests/index.rs:1]: Yes, we need it for the `to_args()` function call in one of the tests.

File: tests/index.rs (hunk 25 above)
Comment (raphjaph) [reply tests/index.rs:40]: The way the our mock bitcoin core instance is written I do not have access to the Basic Authentication http header. So I can't know what credentials are passed in.

File: tests/server.rs (hunk 26 above)
Comment (raphjaph) [reply tests/server.rs:303]: see above

**Review by casey — APPROVED** [review]
